Firestore/core/src/firebase/firestore/local/query_engine.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_QUERY_ENGINE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_QUERY_ENGINE_H_
+
+#include "Firestore/core/src/firebase/firestore/core/query.h"
+#include "Firestore/core/src/firebase/firestore/local/local_documents_view.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/document_map.h"
+#include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+/**
+ * Represents a query engine capable of performing queries over the local
+ * document cache. You must call setLocalDocumentsView() before using.
+ */
+class QueryEngine {
+ public:
+  virtual ~QueryEngine() {
+  }
+
+  /** Sets the document view to query against. */
+  virtual void SetLocalDocumentsView(LocalDocumentsView* local_documents) = 0;
+
+  /** Returns all local documents matching the specified query. */
+  virtual model::DocumentMap GetDocumentsMatchingQuery(
+      core::Query query,
```

Comment:
Passing arguments like this makes copies. These likely should be const references (`const core::Query&`).

Reply:
Passing by const reference is even cheaper; much cheaper.

`DocumentKeySet` and similar types are documented as "cheap" to copy in that they're O(1) to copy when the expectation would be O(n) for a deep copy. Internally they keep a shared_ptrs to their immutable representation and copies are really just reference count increments.

However, this is still an atomic integer operation which has a nontrivial cost that can easily be avoided.

As a general rule:
  * Pass by value
    * Primitives (e.g. ints, doubles)
    * Very small, trivial structs/classes (e.g. `string_view` or `Timestamp`)
    * Move-only types (e.g. `std::unique_ptr`)
    * When you're going to make a copy anyway. This gives the caller the opportunity to give away their copy with std::move (e.g. constructor arguments that will be assigned to members)
  * Otherwise, pass by const reference

Reply:
Updated the PR to pass by const reference. Thanks for providing the guidelines, this should take some (maybe even all) of the guesswork out when it comes to picking the right argument type.